File: typescript/src/bin/compress.ts
#!/usr/bin/env node
/**
 * CLI tool for compressing text using compression-prompt
 */

import * as fs from 'fs';
import * as process from 'process';
import { Compressor, CompressionError, OutputFormat } from '../compressor';

interface CliArgs {
  inputFile?: string;
  outputFile?: string;
  ratio: number;
  stats: boolean;
  help: boolean;
  format: 'text' | 'png' | 'jpeg';
  jpegQuality: number;
}

function printUsage(): void {
  console.log(`Usage: compress [OPTIONS] [INPUT_FILE]

Options:
  -r, --ratio <RATIO>        Compression ratio (0.0-1.0, default: 0.5)
  -o, --output <FILE>        Output file (default: stdout for text, compressed.png for image)
  -f, --format <FORMAT>      Output format: text, png, jpeg (default: text)
  --jpeg-quality <QUALITY>   JPEG quality (1-100, default: 85)
  -s, --stats                Show compression statistics
  -h, --help                 Show this help message

Examples:
  compress input.txt                          # Compress to stdout
  compress -r 0.7 input.txt                   # Conservative (70%)
  compress -r 0.3 input.txt                   # Aggressive (30%)
  compress -f png -o out.png input.txt        # Generate PNG image
  compress -f jpeg --jpeg-quality 90 in.txt   # Generate high-quality JPEG
  cat input.txt | compress                    # Read from stdin
  compress -s input.txt                       # Show statistics
`);
}

function parseArgs(argv: string[]): CliArgs {
  const args: CliArgs = {
    ratio: 0.5,
    stats: false,
    help: false,
    format: 'text',
    jpegQuality: 85,
  };

  for (let i = 2; i < argv.length; i++) {
    const arg = argv[i];

    switch (arg) {
      case '-h':
      case '--help':
        args.help = true;
        break;

      case '-r':
      case '--ratio':
        i++;
        if (i >= argv.length) {
          throw new Error('Missing value for --ratio');
        }
        args.ratio = parseFloat(argv[i]);
        if (isNaN(args.ratio) || args.ratio < 0 || args.ratio > 1) {
          throw new Error('Ratio must be between 0.0 and 1.0');
        }
        break;

      case '-o':
      case '--output':
        i++;
        if (i >= argv.length) {
          throw new Error('Missing value for --output');
        }
        args.outputFile = argv[i];
        break;

      case '-s':
      case '--stats':
        args.stats = true;
        break;

      case '-f':
      case '--format': {
        i++;
        if (i >= argv.length) {
          throw new Error('Missing value for --format');
        }
        const format = argv[i].toLowerCase();
        if (!['text', 'png', 'jpeg', 'jpg'].includes(format)) {
          throw new Error(`Invalid format '${format}'. Use: text, png, or jpeg`);
        }
        args.format = (format === 'jpg' ? 'jpeg' : format) as 'text' | 'png' | 'jpeg';
        break;
      }

      case '--jpeg-quality':
        i++;
        if (i >= argv.length) {
          throw new Error('Missing value for --jpeg-quality');
        }
        args.jpegQuality = parseInt(argv[i]);
        if (isNaN(args.jpegQuality) || args.jpegQuality < 1 || args.jpegQuality > 100) {
          throw new Error('JPEG quality must be between 1 and 100');
        }
        break;

      default:
        if (arg.startsWith('-')) {
          throw new Error(`Unknown option: ${arg}`);
        }
        args.inputFile = arg;
        break;
    }
  }

  return args;
}

async function readInput(args: CliArgs): Promise<string> {
  if (args.inputFile) {
    return fs.readFileSync(args.inputFile, 'utf8');
  } else {
    // Read from stdin
    const chunks: Buffer[] = [];
    for await (const chunk of process.stdin) {
      chunks.push(chunk);
    }
    return Buffer.concat(chunks).toString('utf8');
  }
}

function writeOutput(args: CliArgs, result: any): void {
  if (args.format === 'text') {
    // Text output
    if (args.outputFile) {
      fs.writeFileSync(args.outputFile, result.compressed, 'utf8');
      if (args.stats) {
        console.error(`Text saved to: ${args.outputFile}`);
      }
    } else {
      process.stdout.write(result.compressed);
    }
  } else {
    // Image output
    if (!result.imageData) {
      throw new Error('Image generation failed');
    }

    const outputPath =
      args.outputFile || (args.format === 'jpeg' ? 'compressed.jpg' : 'compressed.png');

    fs.writeFileSync(outputPath, result.imageData);

    if (args.stats) {
      console.error(`Image saved to: ${outputPath}`);
      console.error(
        `Image size: ${Math.floor(result.imageData.length / 1024)} KB (${result.imageData.length} bytes)`
      );
      console.error(`Dimensions: 1024x1024`);

      // Verify format
      const isPng = result.imageData[0] === 137 && result.imageData[1] === 80;
      const isJpeg = result.imageData[0] === 0xff && result.imageData[1] === 0xd8;

      if (isPng) {
        console.error('Format: PNG ✓');
      } else if (isJpeg) {
        console.error('Format: JPEG ✓');
      }

      // Text saved separately
      const textPath = outputPath.replace(/\.(png|jpg|jpeg)$/i, '.txt');
      fs.writeFileSync(textPath, result.compressed);
      console.error(`Compressed text saved to: ${textPath}`);
    } else {
      console.error(`Saved to: ${outputPath}`);
    }
  }
}

async function main() {
  try {
    const args = parseArgs(process.argv);

    if (args.help) {
      printUsage();
      process.exit(0);
    }

    const input = await readInput(args);

    const compressor = new Compressor({
      targetRatio: args.ratio,
      minInputBytes: 100,
      minInputTokens: 10,
    });

    // Determine output format
    const format = args.format === 'text' ? OutputFormat.TEXT : OutputFormat.IMAGE;

    const formatOptions = {
      imageFormat: args.format === 'jpeg' ? ('jpeg' as const) : ('png' as const),
      jpegQuality: args.jpegQuality,
    };

    const result = compressor.compressWithFormat(input, format, formatOptions);

    if (args.stats) {
      console.error('Compression Statistics:');
      console.error(`  Original tokens:   ${result.originalTokens}`);
      console.error(`  Compressed tokens: ${result.compressedTokens}`);
      console.error(`  Tokens removed:    ${result.tokensRemoved}`);
      console.error(`  Compression ratio: ${((1.0 - result.compressionRatio) * 100).toFixed(1)}%`);
      console.error(`  Target ratio:      ${((1.0 - args.ratio) * 100).toFixed(1)}%`);
      console.error('');
    }

    writeOutput(args, result);
  } catch (error) {
    if (error instanceof CompressionError) {
      console.error(`Compression error: ${error.message}`);
      process.exit(1);
    } else if (error instanceof Error) {
      console.error(`Error: ${error.message}`);
      printUsage();
      process.exit(1);
    } else {
      console.error('Unknown error occurred');
      process.exit(1);
    }
  }
}

main();

File: rust/src/bin/compress.rs
//! CLI tool for compressing text using compression-prompt

use compression_prompt::{Compressor, CompressorConfig, OutputFormat, StatisticalFilterConfig};
use std::env;
use std::fs;
use std::io::{self, Read, Write};
use std::path::PathBuf;
use std::process;

fn print_usage() {
    eprintln!("Usage: compress [OPTIONS] [INPUT_FILE]");
    eprintln!();
    eprintln!("Options:");
    eprintln!("  -r, --ratio <RATIO>      Compression ratio (0.0-1.0, default: 0.5)");
    eprintln!("  -o, --output <FILE>      Output file (default: stdout)");
    eprintln!("  -f, --format <FORMAT>    Output format: text, png, jpeg (default: text)");
    eprintln!("  -q, --quality <QUALITY>  JPEG quality 1-100 (default: 85, only for jpeg)");
    eprintln!("  -s, --stats              Show compression statistics");
    eprintln!("  -h, --help               Show this help message");
    eprintln!();
    eprintln!("Examples:");
    eprintln!("  compress input.txt                        # Compress to stdout");
    eprintln!("  compress -r 0.7 input.txt                 # Conservative (70%)");
    eprintln!("  compress -r 0.3 input.txt                 # Aggressive (30%)");
    eprintln!("  compress -f png -o out.png input.txt      # Save as PNG image");
    eprintln!("  compress -f jpeg -q 90 -o out.jpg input.txt  # Save as JPEG");
    eprintln!("  cat input.txt | compress                  # Read from stdin");
}

enum OutputFormatType {
    Text,
    Png,
    Jpeg,
}

struct Config {
    input_file: Option<PathBuf>,
    output_file: Option<PathBuf>,
    compression_ratio: f32,
    show_stats: bool,
    output_format: OutputFormatType,
    jpeg_quality: u8,
}

impl Config {
    fn parse_args() -> Result<Self, String> {
        let args: Vec<String> = env::args().collect();

        let mut config = Config {
            input_file: None,
            output_file: None,
            compression_ratio: 0.5,
            show_stats: false,
            output_format: OutputFormatType::Text,
            jpeg_quality: 85,
        };

        let mut i = 1;
        while i < args.len() {
            match args[i].as_str() {
                "-h" | "--help" => {
                    print_usage();
                    process::exit(0);
                }
                "-r" | "--ratio" => {
                    i += 1;
                    if i >= args.len() {
                        return Err("Missing value for --ratio".to_string());
                    }
                    config.compression_ratio = args[i]
                        .parse::<f32>()
                        .map_err(|_| format!("Invalid ratio: {}", args[i]))?;

                    if !(0.0..=1.0).contains(&config.compression_ratio) {
                        return Err("Ratio must be between 0.0 and 1.0".to_string());
                    }
                }
                "-o" | "--output" => {
                    i += 1;
                    if i >= args.len() {
                        return Err("Missing value for --output".to_string());
                    }
                    config.output_file = Some(PathBuf::from(&args[i]));
                }
                "-f" | "--format" => {
                    i += 1;
                    if i >= args.len() {
                        return Err("Missing value for --format".to_string());
                    }
                    config.output_format = match args[i].to_lowercase().as_str() {
                        "text" => OutputFormatType::Text,
                        "png" => OutputFormatType::Png,
                        "jpeg" | "jpg" => OutputFormatType::Jpeg,
                        _ => {
                            return Err(format!(
                                "Invalid format: {} (use: text, png, jpeg)",
                                args[i]
                            ));
                        }
                    };
                }
                "-q" | "--quality" => {
                    i += 1;
                    if i >= args.len() {
                        return Err("Missing value for --quality".to_string());
                    }
                    config.jpeg_quality = args[i]
                        .parse::<u8>()
                        .map_err(|_| format!("Invalid quality: {}", args[i]))?;

                    if !(1..=100).contains(&config.jpeg_quality) {
                        return Err("Quality must be between 1 and 100".to_string());
                    }
                }
                "-s" | "--stats" => {
                    config.show_stats = true;
                }
                arg if arg.starts_with('-') => {
                    return Err(format!("Unknown option: {}", arg));
                }
                arg => {
                    config.input_file = Some(PathBuf::from(arg));
                }
            }
            i += 1;
        }

        Ok(config)
    }
}

fn read_input(config: &Config) -> io::Result<String> {
    if let Some(ref input_file) = config.input_file {
        fs::read_to_string(input_file)
    } else {
        let mut buffer = String::new();
        io::stdin().read_to_string(&mut buffer)?;
        Ok(buffer)
    }
}

fn write_output(config: &Config, text: &str) -> io::Result<()> {
    if let Some(ref output_file) = config.output_file {
        fs::write(output_file, text)
    } else {
        io::stdout().write_all(text.as_bytes())?;
        io::stdout().flush()
    }
}

fn main() {
    let config = match Config::parse_args() {
        Ok(c) => c,
        Err(e) => {
            eprintln!("Error: {}", e);
            eprintln!();
            print_usage();
            process::exit(1);
        }
    };

    // Read input
    let input = match read_input(&config) {
        Ok(text) => text,
        Err(e) => {
            eprintln!("Error reading input: {}", e);
            process::exit(1);
        }
    };

    // Configure compressor
    let compressor_config = CompressorConfig {
        target_ratio: config.compression_ratio,
        min_input_bytes: 100, // Lower threshold for CLI
        min_input_tokens: 10,
    };

    let filter_config = StatisticalFilterConfig {
        compression_ratio: config.compression_ratio,
        ..Default::default()
    };

    let compressor = Compressor::with_filter_config(compressor_config, filter_config);

    // Determine output format
    let output_format = match config.output_format {
        OutputFormatType::Text => OutputFormat::Text,
        OutputFormatType::Png | OutputFormatType::Jpeg => {
            #[cfg(not(feature = "image"))]
            {
                eprintln!("Error: Image output requires the 'image' feature");
                eprintln!("Rebuild with: cargo build --features image");
                process::exit(1);
            }
            #[cfg(feature = "image")]
            OutputFormat::Image
        }
    };

    // Compress
    let result = match compressor.compress_with_format(&input, output_format) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Compression error: {}", e);
            process::exit(1);
        }
    };

    // Show stats if requested
    if config.show_stats {
        eprintln!("Compression Statistics:");
        eprintln!("  Original tokens:   {}", result.original_tokens);
        eprintln!("  Compressed tokens: {}", result.compressed_tokens);
        eprintln!("  Tokens removed:    {}", result.tokens_removed);
        eprintln!(
            "  Compression ratio: {:.1}%",
            (1.0 - result.compression_ratio) * 100.0
        );
        eprintln!(
            "  Target ratio:      {:.1}%",
            (1.0 - config.compression_ratio) * 100.0
        );
        eprintln!();
    }

    // Write output based on format
    match config.output_format {
        OutputFormatType::Text => {
            if let Err(e) = write_output(&config, &result.compressed) {
                eprintln!("Error writing output: {}", e);
                process::exit(1);
            }
        }
        OutputFormatType::Png => {
            #[cfg(feature = "image")]
            {
                if let Some(image_data) = result.image_data {
                    let output_path = config
                        .output_file
                        .as_ref()
                        .map(|p| p.to_owned())
                        .unwrap_or_else(|| PathBuf::from("compressed.png"));

                    if let Err(e) = fs::write(&output_path, image_data) {
                        eprintln!("Error writing PNG: {}", e);
                        process::exit(1);
                    }

                    if config.show_stats {
                        eprintln!("  Output saved to: {}", output_path.display());
                    }
                } else {
                    eprintln!("Error: Failed to generate PNG image");
                    process::exit(1);
                }
            }
        }
        OutputFormatType::Jpeg => {
            #[cfg(feature = "image")]
            {
                use compression_prompt::ImageRenderer;

                let renderer = ImageRenderer::default();
                match renderer.render_to_jpeg(&result.compressed, config.jpeg_quality) {
                    Ok(jpeg_data) => {
                        let output_path = config
                            .output_file
                            .as_ref()
                            .map(|p| p.to_owned())
                            .unwrap_or_else(|| PathBuf::from("compressed.jpg"));

                        if let Err(e) = fs::write(&output_path, jpeg_data) {
                            eprintln!("Error writing JPEG: {}", e);
                            process::exit(1);
                        }

                        if config.show_stats {
                            eprintln!("  Output saved to: {}", output_path.display());
                        }
                    }
                    Err(e) => {
                        eprintln!("Error generating JPEG: {}", e);
                        process::exit(1);
                    }
                }
            }
        }
    }
}
